README.md
# README

Info and documentation for binance-trader

## About

In its current state, binance-trader is a crypto trading application configured against binanceus via the ccxt API.

NOTE: This is a work in progress. Master branch will generally have working & functional POC code, but v1.0.0 will be the first fully functional application.

## Technology

Built as a command line node.js application.

### Stack

-   Node

### Dependencies

-   [ccxt](https://github.com/ccxt/ccxt)
-   [dotenv](https://github.com/motdotla/dotenv)
-   [minimist](https://github.com/substack/minimist)

### Dev Dependencies

-   [eslint](https://github.com/eslint/eslint)
-   [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier)
-   [eslint-plugin-node](https://github.com/mysticatea/eslint-plugin-node)
-   [eslint-plugin-prettier](https://github.com/prettier/eslint-plugin-prettier)
-   [prettier](https://github.com/prettier/prettier)
-   [husky](https://github.com/typicode/husky)
-   [colors](https://github.com/Marak/colors.js)

## Setup

Pretty basic, clone and run the following in your terminal:

```
npm i
touch .env
```

To utilize the [Binance US](https://www.binance.us/en/home) API, add the following into your `.env` file

```
API_KEY={api key in binance)
API_SECRET={api secret, only shown at creation of api key}
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.12.1] - 2021-03-17

### Updated

-   terminal readout to include colors
-   small adjustments to the pocLive variables

## [0.12.0] - 2021-03-13

### Added

-   Moving average and adjust logic to utilize it

### Updated

-   buy & sell to utilize an owned boolean
-   live poc to no longer log theory data

## [0.11.2] - 2021-03-08

### Updated

-   Initialization to use historica data or "interval" amount of tickers
-   refactored buy/sell to be exponential based on account asset & base amounts

## [0.11.1] - 2021-03-08

### Updated

-   coinData.previous to include 5 previous ticks
-   coinData to capture volume, v from OHLCV endpoint
-   ping to be just be current tick

## [0.11.0] - 2021-03-08

### Added

-   Created pocLive which buys/sells via API and stores data

## [0.10.0] - 2021-03-01

### Added

-   data collection functionality
-   doji corrections to prevent false flips

## [0.9.0] - 2021-03-01

### Added

-   Completely rebuilt histoic poc to run against live data
-   Commented extensively and adde jsDocs to functions

## [0.8.0] - 2021-02-28

### Added

-   logic to read through fetchOHLCV API array of last 500 candles
-   new poc to run against historic data
-   ability to store candles and factor on specific intervals

### Updated

-   npm scripts to take interval arg, 5 being the ideal interval after numerous tests
-   lots of other things...

## [0.7.0] - 2021-02-28

### Added

-   live price data pulled from fetchOHLCV endpoint, refactored data accordingly
-   rebuilt heinkin ashi logic and moved it out of the object config
-   hollowCandle boolean and reconfigured volatility logic accordingly

### Updated

-   poc to take CLI arguments instead of hardcoded vars
-   cleaned up function naming

## [0.6.0] - 2021-02-26

### Added

-   ogject literals to config vs classes
-   heinkin ashi object to tick (but built it wrong...)
-   cuurent and previous as objects and created a "intervalTick" object

### Updated

-   pulled more dynamic info from fetchTick endpoint

## [0.5.3] - 2021-02-23

### Added

-   minimist package
-   scripts for dynamic pairs and common pairs

### Updated

-   poc to take CLI arguments instead of hardcoded vars

## [0.5.2] - 2021-02-23

### Added

-   additional binanceClient.fetchBalance() data instead of config variables

### Removed

-   a bunch junk no longerd needed

## [0.5.1] - 2021-02-22

### Added

-   binanceClient.fetchTicker() to the file in place of coinGecko accross all files

### Removed

-   coingecko and axios dependencies

## [0.5.0] - 2021-02-22

### Added

-   buy & sell functionality from binanceClient.createOrder() method
-   binanceClient.fetchTicker() to the file in place of coinGecko

## [0.4.2] - 2021-02-21

### Added

-   createDataCollectionJSON function to create a new file with the date stamp and coinID/currency

### Updated

-   dataCollector to include a H:M:S timestamp for each ping

## [0.4.1] - 2021-02-21

### Updated

-   CoinData class to have coinID & currency
-   dataCollector in index.js to include account and coinData

### Removed

-   older data collections from previous object shape
-   /helpers/dataCollector.js function
-   /algorithims/volatilityPOC.js

## [0.4.0] - 2021-02-21

### Added

-   combination of POC and dataCollector to index, works against live charts
-   Account and CoinData class objects for tracking live state

### Updated

-   POC and other helper files

## [0.3.1] - 2021-02-20

### Updated

-   algorithm to account for price duplicates
-   directories for helpers and alogrithms
-   data objects

## [0.3.0] - 2021-02-20

### Added

-   volatility logic to index.js to work against captured data
-   more captured data from 8hrs of dogecoin

### Updated

-   trimmed data objects to just be data.price

## [0.2.0] - 2021-02-19

### Added

-   dataCollector.js to capture live market prices
-   /data directory with raw and trimmed price data

### Updated

-   demoTrader to use dynamic API function

## [0.1.5] - 2021-02-17

### Updated

-   coingGecko urls to be a dynamic helper function

## [0.1.4] - 2021-02-16

### Added

-   Husky with precommit linting

## [0.1.3] - 2021-02-16

### Added

-   eslint configuration for project
-   prettier configuration for project

### Updated

-   linted all .js and .json files

## [0.1.2] - 2021-02-16

### Updated

-   CHANGELOG.md - created formatting
-   README.md - simple purpose and instructions

## [0.1.1] - 2021-02-15

### Added

-   /helpers
-   compoundCalc.js simple function for calculating compound interest
-   volatility prototype POC with mock data

## [0.1.0] - 2021-02-15

### Added

-   Working demo and API connections

[0.12.1]: https://github.com/JesseChamberlain/binance-trader/compare/v0.12.0...v0.12.1
[0.12.0]: https://github.com/JesseChamberlain/binance-trader/compare/v0.11.2...v0.12.0
[0.11.2]: https://github.com/JesseChamberlain/binance-trader/compare/v0.11.1...v0.11.2
[0.11.1]: https://github.com/JesseChamberlain/binance-trader/compare/v0.11.0...v0.11.1
[0.11.0]: https://github.com/JesseChamberlain/binance-trader/compare/v0.10.0...v0.11.0
[0.10.0]: https://github.com/JesseChamberlain/binance-trader/compare/v0.9.0...v0.10.0
[0.9.0]: https://github.com/JesseChamberlain/binance-trader/compare/v0.8.0...v0.9.0
[0.8.0]: https://github.com/JesseChamberlain/binance-trader/compare/v0.7.0...v0.8.0
[0.7.0]: https://github.com/JesseChamberlain/binance-trader/compare/v0.6.0...v0.7.0
[0.6.0]: https://github.com/JesseChamberlain/binance-trader/compare/v0.5.3...v0.6.0
[0.5.3]: https://github.com/JesseChamberlain/binance-trader/compare/v0.5.2...v0.5.3
[0.5.2]: https://github.com/JesseChamberlain/binance-trader/compare/v0.5.1...v0.5.2

algorithms/pocLive.js
/* eslint-disable node/no-unpublished-require */
require('dotenv').config();
const ccxt = require('ccxt');
const parseArgs = require('minimist');
const data = require('../helpers/dataCollectors');
// eslint-disable-next-line no-unused-vars
const colors = require('colors');

/**
 * Creates a current moving avg from h.a. close (ohlc/4) of current & previous ticks
 * @param {object} coinData - Object of current and previous coin data
 * @return {float} - returns moving average as a float
 */
const factorMovingAvg = (coinData) => {
    const { current, previous } = coinData;

    // create an array of all current and previous closes
    let closes = [];
    closes.push(current.close);
    previous.forEach((tick) => {
        closes.push(tick.close);
    });

    // take the average of all closes
    const movingAvg = closes.reduce((a, b) => a + b, 0) / closes.length;

    return movingAvg;
};

/**
 * Takes the last OHLCV response and creates a heikin ashi candle
 * @param {array} lastOHLCV - latest OHLCV response [ t, o, h, l, c, v]
 * @param {float} prevOpen - previous heikin ashi candle open
 * @param {float} prevClose - previous heikin ashi candle close
 * @param {boolean} prevHollowCandle - previous heikin ashi candle trend
 * @return {object} - returns a heikin ashi candle object
 */
const createHeikinAshiTick = (
    lastOHLCV,
    prevOpen,
    prevClose,
    prevHollowCandle,
    owned
) => {
    const [
        lastTime,
        lastOpen,
        lastHigh,
        lastLow,
        lastClose,
        lastVol,
    ] = lastOHLCV;
    const open = (prevOpen + prevClose) / 2;
    const close = (lastOpen + lastClose + lastHigh + lastLow) / 4;
    const high = Math.max(open, close, lastHigh);
    const low = Math.min(open, close, lastLow);
    const candleAvg = (open + close) / 2;
    const candleSpread = (close - open) / candleAvg;
    const shadowAvg = (high + low) / 2;
    const shadowSpread = (high - low) / shadowAvg;

    // TODO: this should really be a separate function
    const isCandleHollow = () => {
        if (candleAvg < shadowAvg) {
            return true; // indicates upward trend
        } else if (candleAvg > shadowAvg) {
            return false; // indicates downward trend
        } else {
            console.log('djoi, prevented false flip'.brightYellow);
            return prevHollowCandle; // djoi, prevent false flip
        }
        // if (candleAvg < shadowAvg) {
        //     return true; // indicates upward trend
        // } else if (candleAvg > shadowAvg) {
        //     if (prevHollowCandle && Math.abs(candleSpread) < 0.001) {
        //         console.log('almost djoi, prevented false flip'.brightYellow);
        //         return true; // almost djoi, prevent false flip
        //     } else {
        //         return false; // indicates downward trend
        //     }
        // } else {
        //     console.log('djoi, prevented false flip'.brightYellow);
        //     return prevHollowCandle; // djoi, prevent false flip
        // }
    };

    let tick = {
        time: lastTime,
        price: lastClose,
        movingAvg: 0,
        open: open,
        close: close,
        high: high,
        low: low,
        volume: lastVol,
        owned: owned,
        hollowCandle: isCandleHollow(),
        candleAvg: candleAvg,
        candleSpread: candleSpread,
        shadowAvg: shadowAvg,
        shadowSpread: shadowSpread,
    };

    return tick;
};

/**
 * Initializes account and coinData objects
 * @param {object} coinData - Object of current and previous coin data
 * @param {Class} binanceClient - ccxt.binance client for API requests
 * @param {string} symbol - market pair symbol, ie: BTC/USDT
 */
const initialize = async (
    coinData,
    binanceClient,
    symbol,
    interval,
    storage
) => {
    // response: array of last 500 OHLCVs [[ t, o, h, l, c, v], ...]
    const resOHLCV = await binanceClient.fetchOHLCV(symbol);

    // use second to last OHLCV, last OHLCV might be incomplete
    // Pre-load storage with (interval - 1) amount of OHLCVs
    let index = interval;
    while (index != 1) {
        const lastOHLCV = resOHLCV[resOHLCV.length - index];
        storage.timestamp = lastOHLCV[0];
        storage.o.push(lastOHLCV[1]);
        storage.h.push(lastOHLCV[2]);
        storage.l.push(lastOHLCV[3]);
        storage.c.push(lastOHLCV[4]);
        storage.v.push(lastOHLCV[5]);
        --index;
    }

    console.log('storage initialized');

    // TODO: need to have at least the first previous initialized
    // could possible load all 5 previous to factor initial trending
    const setOHLCV = resOHLCV[resOHLCV.length - interval];
    coinData.previous.push(
        createHeikinAshiTick(
            setOHLCV, // pass the full array for creation
            setOHLCV[1], // open, initializes previous.open
            setOHLCV[4], // close, initializes previous.close
            true, // initialize true for starting hollowCandle state
            false // owned always initialized to false
        )
    );
};

/**
 * Interval function that runs via setInterval every X milliseconds
 * The function polls for data, sets it to storage, and factors on Y intervals
 * @param {object} coinData - Object of current and previous coin data
 * @param {Class} binanceClient - ccxt.binance client for API requests
 * @param {string} symbol - market pair symbol, ie: BTC/USDT
 * @param {integer} interval - interval to create candles and factor logic
 * @param {object} storage - Object of arrays to collect OHLC values
 * @param {string} filePath - path to created JSON file
 */
const tick = async (
    coinData,
    binanceClient,
    symbol,
    interval,
    storage,
    filePath
) => {
    // response: array of last 500 OHLCVs [[ t, o, h, l, c, v], ...]
    const resOHLCV = await binanceClient.fetchOHLCV(symbol);

    // last candle might be incomplete, use 2nd to last
    const lastOHLCV = resOHLCV[resOHLCV.length - 2];

    // if a new timestamp appears, add the OHLCV to storage
    if (storage.timestamp != lastOHLCV[0]) {
        storage.timestamp = lastOHLCV[0];
        storage.o.push(lastOHLCV[1]);
        storage.h.push(lastOHLCV[2]);
        storage.l.push(lastOHLCV[3]);
        storage.c.push(lastOHLCV[4]);
        storage.v.push(lastOHLCV[5]);

        // once an invterval number of items are in storage, create a H.A. tick
        if (storage.o.length == interval) {
            const t = storage.timestamp;
            const o = storage.o[0];
            const h = Math.max(...storage.h);
            const l = Math.min(...storage.l);
            const c = storage.c[storage.o.length - 1];
            const v = storage.v.reduce((a, b) => a + b, 0);
            const storageToFactor = [t, o, h, l, c, v];
            const { open, close, hollowCandle, owned } = coinData.previous[0];

            // passes the selective data above to create a H.A. tick
            let intervalTick = createHeikinAshiTick(
                storageToFactor,
                open,
                close,
                hollowCandle,
                owned
            );
            coinData.current = intervalTick;
            coinData.current.movingAvg = factorMovingAvg(coinData);

            // order
            const binanceTicker = await binanceClient.fetchTicker(symbol);

            // request account balance to set buy/sell
            const { current, previous, coinID, currency } = coinData;
            const accountBalance = await binanceClient.fetchBalance();
            const assetAvailable = accountBalance.free[coinID];
            const baseAvailable = accountBalance.free[currency];
            const amountToBuy =
                Math.round(
                    ((baseAvailable - 1000) / previous[0].price) * 10000
                ) / 10000;

            // Primary conditionals to decide to buy or sell
            // buy
            if (
                current.hollowCandle &&
                !current.owned &&
                current.candleAvg > current.movingAvg
            ) {
                await binanceClient.createOrder(
                    symbol,
                    'market',
                    'buy',
                    amountToBuy,
                    binanceTicker
                );
                current.owned = true;
                console.log('Bought asset!'.brightGreen);
            }

            // sell
            if (
                !current.hollowCandle &&
                current.owned &&
                assetAvailable > 1 &&
                current.candleAvg < current.movingAvg
            ) {
                await binanceClient.createOrder(
                    symbol,
                    'market',
                    'sell',
                    assetAvailable,
                    binanceTicker
                );
                current.owned = false;
                console.log('Sold asset!'.brightRed);
            }

            // request datetime from fetchTicker endpoint
            const resTicker = await binanceClient.fetchTicker(symbol);

            // write current state to the json file
            const ping = {
                time: resTicker.datetime,
                current: coinData.current,
            };
            data.collect(ping, filePath);

            // terminal logging
            console.log(resTicker.datetime);
            console.log(symbol, lastOHLCV[4]);

            // add current tick to begining of coinData.previous array
            coinData.previous.unshift(coinData.current);

            // purge oldest tick, but keep 5 tickets for trend data.
            if (coinData.previous.length > 4) {
                coinData.previous.pop();
            }

            // reset storage and set current to previous
            storage.o = [];
            storage.h = [];
            storage.l = [];
            storage.c = [];
            storage.v = [];
        }
    }
};

// Primary runner
const run = () => {
    const args = parseArgs(process.argv.slice(2)); // command line arguments
    const asset = `${args.ASSET}`; // Coin asset
    const base = `${args.BASE}`; // Base coin for asset (USD, USDT, BTC usually)
    const interval = `${args.INTERVAL}`; // interval to factor candles (15-20 ideal)
    const symbol = `${asset}/${base}`; // symbol used by API calls (BTC/USD, ETH/USD)
    console.log(args);

    // Instantiate binance client using the US binance API
    const binanceClient = new ccxt.binanceus({
        apiKey: process.env.API_KEY,
        secret: process.env.API_SECRET,
    });

    // CoinData object for storing and mutating current state of coin price
    let coinData = {
        coinID: asset,
        currency: base,
        current: {},
        previous: [],
    };

    // Storage object of arrays for collecting intervals of OHLCV ticks
    let storage = {
        timestamp: [],
        o: [],
        h: [],
        l: [],
        c: [],
        v: [],
    };

    // Create filePath for writing data
    const filePath = data.createCollectionJSON(coinData);

    // Initialize the account and coin data
    initialize(coinData, binanceClient, symbol, interval, storage);

    // Poll every 5000 milliseconds and run tick()
    setInterval(
        tick,
        5000,
        coinData,
        binanceClient,
        symbol,
        interval,
        storage,
        filePath
    );
};

// Fire it up
run();
